pad_server/src/actuator.h

```diff
@@ -42,6 +43,7 @@ typedef struct actuator {
 int actuator_on(actuator_t *act);
 int actuator_off(actuator_t *act);
 void actuator_init(actuator_t *act, uint8_t id, actuate_f on, actuate_f off, void *priv);
+int actuator_set(padstate_t *state, uint8_t id, uint8_t new_state);
```

Comment:
I'm trying to see the vision for the architecture here:

The ` actuator_set` function gets the actuator state, checks if the new state is different than the current state and calls either the on and off function.
Internally the on and off functions call `padstate_actuate`, in the future they will have additional logic for each actuator, that's why an array of the `actuator_t` type is needed.

I don't get this part
> and whether or not the request is OK/DENIED.

Does this mean the validity check for the actuation request needs to be checked in the on/off functions? I'm confused.

Reply:
For the architecture, the goal is that each actuator will provide only two functions: on and off. Those are what are used to control the actuator.

Since the command communication spec allows setting of the actuator state itself, I think it's fair to have an `actuator_set` command like you implemented. However, this shouldn't be coupled with the logic for arming, it should just call either on or off functions based on the requested state.

```c
// Something like this?
int actuator_set(act_t *act, bool requested_state) {
    if (requested_state) {
        return act->on(act);
    }
    return act->off(act);
}
```

Instead, the arming checks should take place in something like the `padstate_actuate` function. This function checks to make sure the actuation is permitted (DENIED/OK), and if it is permitted then it actually calls `actuator_set` on the actuator. All the actuator* commands should blindly execute and not perform arming logic error checking, that will be a calling code responsibility. Basically move the switch statement into another function and have it verify that the actuation would be allowed before performing it

I believe the current implementation has `actuator_set` performing the logic checks. Any functions that are prefixed with `actuator` should, in my opinion, be limited to only operating on the actuator and not requiring knowledge of the arming state. If you have a different idea for the architecture I would be happy to hear it though!

Reply:
Ok I see. So the chain of command would be
controller -> padstate_actuate (does the checks, retrieves the actuator) -> actuator_set -> on (operations with pthread)

If that's the case it's kinda awkward because the other functions starting with padstate lock the state directly, but I can work around that and change the naming here and there.

Reply:
The `padstate_actuate` will have to lock the state to check the arming level, so not really far removed! But yep, that's the chain of command. I don't think `on` does any pthread operations though.